[GTK][WPE] Remove invalid flag from Damage

carlosgcampos opened the pull request:
#### f84fa6a88b80c399fa10ed4b061021cd6c438ab3
<pre>
[GTK][WPE] Remove invalid flag from Damage
<a href="https://bugs.webkit.org/show_bug.cgi?id=290098">https://bugs.webkit.org/show_bug.cgi?id=290098</a>

Reviewed by Alejandro G. Castro.

The invalid flag is very confusing, it&apos;s not clear the difference
between empty and invalid Damage, and it&apos;s used with different meanings
in different places. We can just use std::optional which always means no
damage, so that the way we handle no damage depends on every case.
This patch also tries to avoid copying Damage whenever possible and
removes the method to get a region, since that&apos;s only used by the
FrameDamageHistory and it&apos;s an implementation detail.

* LayoutTests/platform/glib/damage/accelerated-canvas-2d-fillRect.html:
* LayoutTests/platform/glib/damage/animations-blur.html:
* LayoutTests/platform/glib/damage/basic-propagation-001.html:
* LayoutTests/platform/glib/damage/basic-propagation-002.html:
* LayoutTests/platform/glib/damage/canvas-webgl.html:
* LayoutTests/platform/glib/damage/common.js:
* LayoutTests/platform/glib/damage/image-as-layer-src-change.html:
* LayoutTests/platform/glib/damage/image-src-change.html:
* LayoutTests/platform/glib/damage/layer-backdrop-filter.html:
* LayoutTests/platform/glib/damage/layer-color-change.html:
* LayoutTests/platform/glib/damage/layer-downsize-with-movement.html:
* LayoutTests/platform/glib/damage/layer-downsize.html:
* LayoutTests/platform/glib/damage/layer-in-layer-movement.html:
* LayoutTests/platform/glib/damage/layer-movement.html:
* LayoutTests/platform/glib/damage/layer-opacity-change.html:
* LayoutTests/platform/glib/damage/layer-overlaps-with-opacity.html:
* LayoutTests/platform/glib/damage/layer-resize.html:
* LayoutTests/platform/glib/damage/layer-text-change.html:
* LayoutTests/platform/glib/damage/layer-upsize.html:
* LayoutTests/platform/glib/damage/layer-visibility-change.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-arc.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-downsizing.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-fillRect-w-offset.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-fillRect.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-new-canvas.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-path.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-putImageData.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-resizing.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-text.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-upsizing.html:
* LayoutTests/platform/glib/damage/scrolling-container-001.html:
* LayoutTests/platform/glib/damage/scrolling-container-002.html:
* LayoutTests/platform/glib/damage/scrolling-container-003.html:
* LayoutTests/platform/glib/damage/scrolling-container-004.html:
* LayoutTests/platform/glib/damage/scrolling-fullscreen-001.html:
* LayoutTests/platform/glib/damage/scrolling-fullscreen-002.html:
* LayoutTests/platform/glib/damage/scrolling-fullscreen-003.html:
* LayoutTests/platform/glib/damage/storage-and-simplification-algorithm.html:
* LayoutTests/platform/glib/damage/transform-2d-rotate.html:
* LayoutTests/platform/glib/damage/transform-2d-scale.html:
* LayoutTests/platform/glib/damage/transform-2d-skew.html:
* LayoutTests/platform/glib/damage/transform-2d-translate.html:
* LayoutTests/platform/glib/damage/transform-3d-dynamic-addition.html:
* LayoutTests/platform/glib/damage/transform-3d-dynamic-re-addition.html:
* LayoutTests/platform/glib/damage/transform-3d-dynamic-removal.html:
* LayoutTests/platform/glib/damage/transform-3d-rotate-1-axis.html:
* LayoutTests/platform/glib/damage/transform-3d-rotate-3-axes-content-update.html:
* LayoutTests/platform/glib/damage/transform-3d-rotate-3-axes.html:
* LayoutTests/platform/glib/damage/transform-3d-rotate-in-still-rotate.html:
* LayoutTests/platform/glib/damage/transform-3d-scale.html:
* LayoutTests/platform/glib/damage/transform-3d-translate-w-perspective-content-update.html:
* LayoutTests/platform/glib/damage/transform-3d-translate-w-perspective.html:
* LayoutTests/platform/glib/damage/transform-3d-translate.html:
* LayoutTests/platform/glib/damage/video.html:
* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::isEmpty const):
(WebCore::Damage::add):
(WebCore::FrameDamageHistory::damageInformation const):
(WebCore::FrameDamageHistory::addDamage):
(WebCore::operator&lt;&lt;):
(WebCore::Damage::invalid): Deleted.
(WebCore::Damage::region const): Deleted.
(WebCore::Damage::isInvalid const): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapper.h:
(WebCore::TextureMapper::setDamage):
(WebCore::TextureMapper::damage const):
* Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
(WebCore::TextureMapperDamageVisualizer::paintDamage const):
* Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::setDamage):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStore::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::flushCompositingState):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::getFrameDamageHistory const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.h:
(WebKit::AcceleratedSurface::addDamage):
(WebKit::AcceleratedSurface::frameDamage const):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::paintToCurrentGLContext):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::addDamage):
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::addDamage):
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::addDamage):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, Basics)):
(TestWebKitAPI::TEST(Damage, AddDamage)):

Canonical link: <a href="https://commits.webkit.org/292420@main">https://commits.webkit.org/292420@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b66e3257ab3e211f546078c25a4172248ff6a628

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96015 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15629 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5576 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101077 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46524 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98060 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15924 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24061 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73198 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30425 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99018 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11921 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86727 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53535 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11664 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4479 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45859 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81813 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4580 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103103 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23082 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82242 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23333 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82741 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81614 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26210 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3647 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16430 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15448 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23045 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22704 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26184 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24445 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Scony commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
```diff
@@ -274,34 +274,37 @@ void ThreadedCompositor::paintToCurrentGLContext(const TransformationMatrix& mat
+        if (m_frameDamageHistory)
+            m_frameDamageHistory->addDamage(frameDamage);
+
+        const auto& damageSinceLastSurfaceUse = m_surface->addDamage(WTFMove(frameDamage));
```
if `frameDamage` is empty at this point we should mark it full-viewport (add a viewport-size rect into it) before passing here, and even before passing to `m_frameDamageHistory` IMO

Scony replied:
ok, looks like it's handled in AcceleratedSurface now, then nvm

carlosgcampos replied:
I agree, but I didn't want to change the behavior in this PR. We can do this in a follow up.

alexgcastro reviewed (approved):
LGTM!

carlosgcampos committed:
[GTK][WPE] Remove invalid flag from Damage
https://bugs.webkit.org/show_bug.cgi?id=290098

Reviewed by Alejandro G. Castro.

The invalid flag is very confusing, it's not clear the difference
between empty and invalid Damage, and it's used with different meanings
in different places. We can just use std::optional which always means no
damage, so that the way we handle no damage depends on every case.
This patch also tries to avoid copying Damage whenever possible and
removes the method to get a region, since that's only used by the
FrameDamageHistory and it's an implementation detail.

* LayoutTests/platform/glib/damage/accelerated-canvas-2d-fillRect.html:
* LayoutTests/platform/glib/damage/animations-blur.html:
* LayoutTests/platform/glib/damage/basic-propagation-001.html:
* LayoutTests/platform/glib/damage/basic-propagation-002.html:
* LayoutTests/platform/glib/damage/canvas-webgl.html:
* LayoutTests/platform/glib/damage/common.js:
* LayoutTests/platform/glib/damage/image-as-layer-src-change.html:
* LayoutTests/platform/glib/damage/image-src-change.html:
* LayoutTests/platform/glib/damage/layer-backdrop-filter.html:
* LayoutTests/platform/glib/damage/layer-color-change.html:
* LayoutTests/platform/glib/damage/layer-downsize-with-movement.html:
* LayoutTests/platform/glib/damage/layer-downsize.html:
* LayoutTests/platform/glib/damage/layer-in-layer-movement.html:
* LayoutTests/platform/glib/damage/layer-movement.html:
* LayoutTests/platform/glib/damage/layer-opacity-change.html:
* LayoutTests/platform/glib/damage/layer-overlaps-with-opacity.html:
* LayoutTests/platform/glib/damage/layer-resize.html:
* LayoutTests/platform/glib/damage/layer-text-change.html:
* LayoutTests/platform/glib/damage/layer-upsize.html:
* LayoutTests/platform/glib/damage/layer-visibility-change.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-arc.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-downsizing.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-fillRect-w-offset.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-fillRect.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-new-canvas.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-path.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-putImageData.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-resizing.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-text.html:
* LayoutTests/platform/glib/damage/non-accelerated-canvas-2d-upsizing.html:
* LayoutTests/platform/glib/damage/scrolling-container-001.html:
* LayoutTests/platform/glib/damage/scrolling-container-002.html:
* LayoutTests/platform/glib/damage/scrolling-container-003.html:
* LayoutTests/platform/glib/damage/scrolling-container-004.html:
* LayoutTests/platform/glib/damage/scrolling-fullscreen-001.html:
* LayoutTests/platform/glib/damage/scrolling-fullscreen-002.html:
* LayoutTests/platform/glib/damage/scrolling-fullscreen-003.html:
* LayoutTests/platform/glib/damage/storage-and-simplification-algorithm.html:
* LayoutTests/platform/glib/damage/transform-2d-rotate.html:
* LayoutTests/platform/glib/damage/transform-2d-scale.html:
* LayoutTests/platform/glib/damage/transform-2d-skew.html:
* LayoutTests/platform/glib/damage/transform-2d-translate.html:
* LayoutTests/platform/glib/damage/transform-3d-dynamic-addition.html:
* LayoutTests/platform/glib/damage/transform-3d-dynamic-re-addition.html:
* LayoutTests/platform/glib/damage/transform-3d-dynamic-removal.html:
* LayoutTests/platform/glib/damage/transform-3d-rotate-1-axis.html:
* LayoutTests/platform/glib/damage/transform-3d-rotate-3-axes-content-update.html:
* LayoutTests/platform/glib/damage/transform-3d-rotate-3-axes.html:
* LayoutTests/platform/glib/damage/transform-3d-rotate-in-still-rotate.html:
* LayoutTests/platform/glib/damage/transform-3d-scale.html:
* LayoutTests/platform/glib/damage/transform-3d-translate-w-perspective-content-update.html:
* LayoutTests/platform/glib/damage/transform-3d-translate-w-perspective.html:
* LayoutTests/platform/glib/damage/transform-3d-translate.html:
* LayoutTests/platform/glib/damage/video.html:
* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::isEmpty const):
(WebCore::Damage::add):
(WebCore::FrameDamageHistory::damageInformation const):
(WebCore::FrameDamageHistory::addDamage):
(WebCore::operator<<):
(WebCore::Damage::invalid): Deleted.
(WebCore::Damage::region const): Deleted.
(WebCore::Damage::isInvalid const): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapper.h:
(WebCore::TextureMapper::setDamage):
(WebCore::TextureMapper::damage const):
* Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
(WebCore::TextureMapperDamageVisualizer::paintDamage const):
* Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::setDamage):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStore::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::flushCompositingState):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::getFrameDamageHistory const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.h:
(WebKit::AcceleratedSurface::addDamage):
(WebKit::AcceleratedSurface::frameDamage const):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::paintToCurrentGLContext):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::addDamage):
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::addDamage):
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::addDamage):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, Basics)):
(TestWebKitAPI::TEST(Damage, AddDamage)):

Canonical link: https://commits.webkit.org/292420@main
